Add WebTonicClient to miden-client

dagarcia7 opened the pull request:
# Summary
This PR adds the `WebRpcClient`, a WASM-compatible implementation of the miden-client `NodeRpcClient` trait.

The `WebRpcClient` is designed for web environments. The existing `tonic_client` builds an auto-generated `ApiClient` using `tonic_build` with the default `transport` feature enabled. As a result, this auto-generated implementation relies heavily on `tokio`, which is not compatible with WASM. This poses a challenge for web-based applications that need to interact with the Miden node.

To address this, the `WebRpcClient` has been developed to provide the same functionality but in a web-compatible manner. The key difference is that `WebRpcClient` also generates an `ApiClient` via `tonic_build`, but with the `transport` feature disabled. When used in combination with the [tonic-web-wasm-client crate](https://crates.io/crates/tonic-web-wasm-client), you can wrap the newly generated `ApiClient` and create an `ApiClient` that is compatible with `grpc-web`. 

# Considerations
I'm sure the biggest thing that will stand out to reviewers is the fact that I had to basically duplicate the `tonic_client/domain` folder to `web_tonic_client/domain` as well. As well as the fact that the auto-generated code to come from both `tonic_build` configurations is the same with the exception of the `rpc.rs` file. I took a while to see I could manipulate the `build.rs` file to dump all auto-generated files with the exception of `rpc.rs` to `/src/client/rpc/generated_common`, and dump the respective `rpc.rs` copies needed (one build with transport, the other without) to `src/client/rpc/tonic_client/generated` and `src/client/rpc/web_tonic_client/generated`, respectively. From here, I wanted to also include a step in the `build.rs` file to apply some post-processing to each `rpc.rs` file so the imports within the auto-generated code could properly point to the right objects in the `generated_common` folder. 

This proved to be a bit difficult, however. Couldn't get it working after a few hours, so decided to put this PR up with the duplication. Happy to discuss alternate approaches, or to see if we can get this working in the follow-up PR.

dagarcia7 commented on Cargo.toml:
```diff
@@ -53,14 +54,18 @@ rusqlite_migration = { version = "1.0", optional = true }
 tracing-subscriber = { version = "0.3" }
 winter-maybe-async = "0.10.0"
 
+[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
+tonic = { version = "0.11", optional = true }
+
 [target.'wasm32-unknown-unknown'.dependencies]
+tonic = { package = "tonic", version="0.11.0", default-features = false, features=["prost","codegen"], optional = true }
```
I wasn't sure of a better way to achieve this. Essentially, the `tonic_client`, needs the default features (which includes the `transport` feature to compile the auto-generated `rpc.rs` code and the `tonic_web_client` needs default features off with these two added in order to compile it's own `rpc.rs` file.

bobbinth replied:
Not sure if this is the best alternative, but we can do something like this:

First, define tonic dependency as:
```toml
tonic = { version = "0.11", default-features = false, optional = true, features = ["prost", "codegen"] }
```
Then, define `tonic` feature as:
```toml
tonic = ["dep:hex", "dep:prost", "dep:tonic", "tonic/transport"]
```

dagarcia7 replied:
Nice! I think this works. Thanks for the suggestion 🎉

dagarcia7 commented on build.rs:
```diff
@@ -32,7 +33,19 @@ fn compile_tonic_client_proto(proto_dir: &Path) -> miette::Result<()> {
+
     // Generate the stub of the user facing server from its proto file
+    tonic_build::configure()
+        .build_transport(false)
+        .build_server(false)
```
These two settings are really the main difference between the files built with this `tonic_build`, and the one below. The documentation here https://crates.io/crates/tonic-web-wasm-client says to do this.

dagarcia7 commented on src/client/rpc/web_tonic_client/domain/accounts.rs:
As described in the PR summary, all of these `domain/*` files are copied from `src/client/rpc/tonic_client/domain/*` with the exception of the paths changed (i.e. `rpc::web_tonic_client::generated` vs. `rpc::tonic_client::generated`). 

Discussed in the PR an approach I attempted to avoid this duplication (as well as the duplication of the auto-generated code), but was unsuccessful.

bobbinth replied:
I don't mind the duplication of generated code, but I'm wondering if we can avoid duplication of the domain objects. One potential way to do it could be to put domain objects into the `src/client/rpc/domain` module. For example, this file would go into `src/client/rpc/domain.accounts.rs` and the top of the file could look something like this:

```rust
use std::fmt::{Debug, Display, Formatter};

use miden_objects::accounts::AccountId;

use crate::errors::RpcConversionError;

#[cfg(feature = "tonic")]
use super::tonic_client::generated::account::AccountId as ProtoAccountId;

#[cfg(feature = "web-tonic")]
use super::web_tonic_client::generated::account::AccountId as ProtoAccountId;
```
I haven't tested whether it'll work - but I think it might.

The only time when it will cause a problem is when `tonic` and `web-tonic` features are enabled at the same time - but we should probably disallow this anyway.

mFragaBA replied:
just in case, I entioned something related [here](https://github.com/0xPolygonMiden/miden-client/pull/409#pullrequestreview-2151825042)

dagarcia7 commented on src/client/rpc/web_tonic_client/generated/account.rs:
As mentioned in the PR summary, all of the generated files with the exception of `rpc.rs` are the same. Also discussed an approach to avoid duplication in this sense in addition to avoiding duplication of the `domain` files, but couldn't get it working.

dagarcia7 commented on src/client/rpc/web_tonic_client/mod.rs:
```diff
@@ -0,0 +1,374 @@
+        Self { endpoint: endpoint.to_string() }
+    }
+
+    pub fn build_api_client(&self) -> ApiClient<Client> {
+        let wasm_client = Client::new(self.endpoint.clone());
+        ApiClient::new(wasm_client)
+    }
```
This is really the main difference between this and the `tonic_client`. This is pulled directly from the `tonic-web-wasm-client` documentation and examples.

dagarcia7 commented on src/client/rpc/web_tonic_client/mod.rs:
```diff
@@ -0,0 +1,374 @@
+        block_num: Option<u32>,
+        include_mmr_proof: bool,
+    ) -> Result<(BlockHeader, Option<MmrProof>), RpcError> {
+        let mut query_client = self.build_api_client();
```
Beyond building the query client slightly differently here than in the `tonic_client`, the rest of the code should be the same. This includes building the request object, calling the appropriate method on the client, and parsing the response to return the appropriate return object. 

If the reviewer thinks it's a good idea, I can look into pulling duplicated logic here into a shared space.

bobbinth replied:
I think this can be done in a follow-up PR. Let's create an issue for this.

dagarcia7 replied:
Sounds great. Created [this issue](https://github.com/0xPolygonMiden/miden-client/issues/415) to track 👍 Put some notes in the issue as well that it's currently blocked on this one.

bobbinth reviewed (commented):
Looks good! Thank you! I left a few comments/suggestions inline.

mFragaBA commented on src/client/rpc/web_tonic_client/domain/mod.rs:
```diff
@@ -0,0 +1,27 @@
+pub mod nullifiers;
+pub mod transactions;
+
+// UTILITIES
+// ================================================================================================
+
+pub trait MissingFieldHelper {
+    fn missing_field(field_name: &'static str) -> RpcConversionError;
+}
+
+impl<T: prost::Message> MissingFieldHelper for T {
+    fn missing_field(field_name: &'static str) -> RpcConversionError {
+        RpcConversionError::MissingFieldInProtobufRepresentation {
+            entity: type_name::<T>(),
+            field_name,
+        }
+    }
+}
```
I think this is not tied to a specific implementation. It could be defined in rpc/mod.rs, couldn't it?

igamigo replied:
This depends on `prost` being enabled which is in turn gated by the `tonic` feature (at least right now).

mFragaBA replied:
can we put it on `rpc/mod.rs` under a `cfg(feature = tonic)` then?

igamigo replied:
Then it would still be tied to a specific implementation, no? But it might be able to be generalized some other way (since the dependency is used by the web tonic feature as well)

mFragaBA replied:
ah, you're right (we could do `cfg(any(feature = tonic, feature = tonic-web))`

dagarcia7 replied:
I think @bobbinth's suggestion [here](https://github.com/0xPolygonMiden/miden-client/pull/409#discussion_r1660074229) cleans this up. I was able to remove the `domain` duplication so this should be the single source of truth again!

mFragaBA reviewed (approved):
Great work! I don't mind the duplication that much for now. Also, if the idea still is to make a wasm crate, we could consider moving the webstore and webtonicclient implementations to that crate if it makes sense (I think that might also remove the need for a wasm feature flag). Left some minor comments/nits

dagarcia7 commented on Cargo.toml:
```diff
@@ -53,14 +54,18 @@ rusqlite_migration = { version = "1.0", optional = true }
+[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
+tonic = { version = "0.11", optional = true }
+
 [target.'wasm32-unknown-unknown'.dependencies]
```
With @bobbinth's suggestion [here](https://github.com/0xPolygonMiden/miden-client/pull/409#discussion_r1660071166), I think I have the code building for both paths without these `[target...]` blocks. After handling the `tonic` dependency with the strategy Bobbin mentioned, I also applied this to `getrandom` and just added this `"getrandom/js"` as a feature to the `web-tonic` feature.

dagarcia7 commented on Cargo.toml:
```diff
@@ -53,14 +54,18 @@ rusqlite_migration = { version = "1.0", optional = true }
 serde_json = { version = "1.0", features = ["raw_value"] }
 tokio = { version = "1.29", features = ["rt-multi-thread", "net", "macros"], optional = true }
-tonic = { version = "0.11", optional = true }
+tonic-web-wasm-client = { version = "0.5", optional = true }
```
Per my comment [above](https://github.com/0xPolygonMiden/miden-client/pull/409#discussion_r1663294834), I currently have these `[target...]` blocks removed so kept this as is for now and it's listed as a dependency to include under the `web-tonic` feature.

dagarcia7 commented:
> Great work! I don't mind the duplication that much for now. Also, if the idea still is to make a wasm crate, we could consider moving the webstore and webtonicclient implementations to that crate if it makes sense (I think that might also remove the need for a wasm feature flag). Left some minor comments/nits

@mFragaBA great point! I think for now it would be best to keep this PR as is and then my next PR planned is the one that will be adding the `web-client` code that references both the `WebStore` and `WebTonicClient`. I think we can have the discussion of it living as its own crate or not there, and I can copy this folder and the `WebStore` code accordingly should we decide to move forward with that route then. 👍

dagarcia7 commented:
I tried running the `make doc` command locally and getting these types of errors

```
error[E0252]: the name `ProtoAccountId` is defined multiple times
 --> crates/rust-client/src/rpc/domain/accounts.rs:9:5
  |
7 | use crate::rpc::tonic_client::generated::account::AccountId as ProtoAccountId;
  |     ------------------------------------------------------------------------- previous import of the type `ProtoAccountId` here
8 | #[cfg(feature = "web-tonic")]
9 | use crate::rpc::web_tonic_client::generated::account::AccountId as ProtoAccountId;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `ProtoAccountId` reimported here
  |
  = note: `ProtoAccountId` must be defined only once in the type namespace of this module
help: you can use `as` to change the binding name of the import
  |
9 | use crate::rpc::web_tonic_client::generated::account::AccountId as OtherProtoAccountId;
```
Is this because the underlying command runs `all-features`? `cargo doc --all-features --keep-going --release -p miden-client`. I could address all of these issues, but then would have to gate all the code in `domain` under different feature flags so don't think we want that. Any suggestions?

bobbinth commented on crates/rust-client/src/rpc/mod.rs:
```diff
@@ -12,13 +12,20 @@ use miden_objects::{
 
 use crate::errors::RpcError;
 
+pub mod domain;
```
I think we should declare this module only if either `tonic` or `web-tonic` features is enabled.

Also, does this need to be `pub mod` or would just `mod` work?

dagarcia7 replied:
I think just `mod` works 👍 and I added the compile flag 💯

bobbinth reviewed (approved):
Looks good! Thank you! I left a couple of small comments inline. After these are addressed, we can merge.

igamigo commented:
> Is this because the underlying command runs all-features? cargo doc --all-features --keep-going --release -p miden-client. I could address all of these issues, but then would have to gate all the code in domain under different feature flags so don't think we want that. Any suggestions?

Yeah, this seems to be the case. As for solutions, I'm not sure what would be the best way to mitigate this right now. I can check tomorrow but some things that come to mind:

- Just generate docs avoiding either `tonic` or `webtonic` for now, and tackling this later with the other code deduplication efforts.
- Keeping the web tonic client on the WASM crate as @mFragaBA suggested.
- Avoiding having similar generated structs on the same module (which would probably entail duplicating more code and gating each of the web/non-web versions at a higher level).

mFragaBA commented on crates/rust-client/Cargo.toml:
```diff
@@ -42,13 +43,12 @@ serde_json = { version = "1.0", features = ["raw_value"] }
 winter-maybe-async = "0.10.0"
-
-[target.'wasm32-unknown-unknown'.dependencies]
 getrandom = { version = "0.2", features = ["js"] }
```
should getrandom be optional? Can it be enabled through the web-tonic feature or is it also used within the store? I don't mind leaving the change for a follow up though

mFragaBA replied:
nvm, I've just seen the last commit changed that

mFragaBA reviewed (approved):
LGTM! @dagarcia7 Lint failed because of fmt.

mFragaBA reviewed (commented):
Oh, I also forgot to mention we should update `CHANGELOG` with this change (I think we also missed it for the WebStore)

dagarcia7 commented:
> LGTM! @dagarcia7 Lint failed because of fmt.

Fixed!

> Oh, I also forgot to mention we should update `CHANGELOG` with this change (I think we also missed it for the WebStore)

Added a message retroactively for WebStore and added one for the WebTonicClient

mFragaBA commented on CHANGELOG.md:
```diff
@@ -1,5 +1,7 @@
 # Changelog
 
+* Added WebTonicClient to the miden-client to support WASM-compatible RPC calls
+* Added WebStore to the miden-client to support WASM-compatible store mechanisms
```
```suggestion
* Added WebTonicClient to the miden-client to support WASM-compatible RPC calls (#409).
* Added WebStore to the miden-client to support WASM-compatible store mechanisms (#401).
```

igamigo reviewed (approved):
LGTM! As Bobbin mentioned, we can tackle any remaining work as part of #415
